spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -327,12 +331,9 @@ private boolean canDeleteUsingMetadata(Expression deleteExpr) {
             .filter(deleteExpr)
             .caseSensitive(caseSensitive)
             .includeColumnStats()
+            .useRef(branch())
             .ignoreResiduals();
 
-    if (branch != null) {
-      scan.useRef(branch);
-    }
```

Comment:
So this was the main issue, the Scan APIs return new Scan objects. It doesn't modify by reference.We either need to do scan = scan.useRef and use that or what I did was just inline it at the time of building

Reply:
I think it would be better to use `scan = scan.useRef(branch)` instead of explicitly passing `"main"`. I'm concerned about relying on `useRef` with an explicit main branch hiding problems with not calling `useRef` at all.